File: solutions/code-carl/ch08_greedy/T11_LC0406_ReconstructQueue.java
/**
 * Practice template for:
 * - LeetCode 406. Queue Reconstruction by Height
 *   https://leetcode.cn/problems/queue-reconstruction-by-height/
 */
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2025/4/7 09:17
 * @comment
 */
public class T11_LC0406_ReconstructQueue {
    public int[][] reconstructQueue(int[][] people) {
        Arrays.sort(people, (a, b) -> {
            int compare = Integer.compare(b[0], a[0]);
            return compare == 0 ? Integer.compare(a[1], b[1]) : compare;
        });

        List<int[]> list = new LinkedList<>();
        for (int[] node : people) {
            list.add(node[1], node);
        }
        return list.toArray(int[][]::new);
    }
}

File: templates/code-carl/ch08_greedy/t11_lc0406_reconstruct_queue.py
"""LeetCode 406. Queue Reconstruction by Height
https://leetcode.cn/problems/queue-reconstruction-by-height/
"""

from typing import List


class Solution:
    def reconstructQueue(self, people: List[List[int]]) -> List[List[int]]:
        pass

if __name__ == "__main__":
    solution = Solution()
    assert solution.reconstructQueue([[7, 0], [4, 4], [7, 1], [5, 0], [6, 1], [5, 2]]) == [[5, 0], [7, 0], [5, 2], [6, 1], [4, 4], [7, 1]]
    assert solution.reconstructQueue([[6, 0], [5, 0], [4, 0], [3, 2], [2, 2], [1, 4]]) == [[4, 0], [5, 0], [2, 2], [3, 2], [1, 4], [6, 0]]
    print("All tests passed.")
